src/StructUtils.hs

```diff
@@ -1,19 +1,33 @@
 module StructUtils where
 
-import Managed
+import Interfaces
 import Obj
 import Polymorphism
 import Types
 
-memberInfo :: TypeEnv -> Ty -> (Ty, String, Ty)
-memberInfo typeEnv memberTy =
-  let refOrNotRefType = if isManaged typeEnv memberTy then RefTy memberTy (VarTy "w") else memberTy -- OBS! The VarTy "w" here is dubious
-   in (refOrNotRefType, if isManaged typeEnv memberTy then "&" else "", FuncTy [refOrNotRefType] StringTy StaticLifetimeTy)
+-- | The 'str'/'prn' functions for primitive types don't take refs, while other types do
+-- so we need to adjust for that when finding and calling them in compound types.
+-- The returned tuple contains ("" || "&", `str function type`).
+memberStrCallingConvention :: String -> TypeEnv -> Env -> Ty -> (String, Ty)
+memberStrCallingConvention strOrPrn typeEnv globalEnv memberTy =
+  if callWithValue
+    then ("", withValueSig)
+    else ("&", withRefSig)
+  where
+    withValueSig = FuncTy [memberTy] StringTy StaticLifetimeTy
+    withRefSig = FuncTy [RefTy memberTy (VarTy "w")] StringTy StaticLifetimeTy -- "w" here is dubious?
+    callWithValue =
+      -- If these interfaces are not implemented we assume ref signature is fine.
+      -- Blitable is required to not accicentally pass a value owned by the struct to
+      -- a non-ref str/prn function implemented on one of its memeber types.
+      let blitable = interfaceImplementedForTy typeEnv globalEnv "blit" (FuncTy [memberTy] memberTy StaticLifetimeTy)
```

Comment:
It seems somewhat unfortunate that we need to check this.

Is there any way for us to make this more general? As in, could we somehow define `str` differently so that the special case is no longer a special case?

Reply:
Yes, we could force `str` to only take refs. I actually think using an interface for properties like this is kinda neat, but if we can avoid it all together that's even better of course.